File: services/api/src/api/database/models/glossary_item.py
from sqlalchemy import BigInteger, Integer, DateTime, ForeignKey, Text, func
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column
from datetime import datetime
from typing import Any

from api.database.db import Base


class GlossaryItem(Base):
    """
    
    
    Parameters:
    
    """
    __tablename__ = "glossary_items"

    # -------------------------
    # Glossary ID / App User ID
    # -------------------------
    id: Mapped[int] = mapped_column(
        BigInteger,
        primary_key=True,
        autoincrement=True,
    )
    user_id: Mapped[int] = mapped_column(
        ForeignKey("app_users.id"),
        index=True,
        nullable=False,
    )

    # -------------------------
    # Clicked word info
    # -------------------------
    form: Mapped[str] = mapped_column(
        Text,
        nullable=False,
    )
    pos_form: Mapped[str] = mapped_column(
        Text,
        nullable=True,
    )
    ipa_form: Mapped[str] = mapped_column(
        Text,
        nullable=True,
    )

    # -------------------------
    # Lemma info
    # -------------------------
    lemma: Mapped[str] = mapped_column(
        Text,
        nullable=True,
    )
    pos_lemma: Mapped[str] = mapped_column(
        Text,
        nullable=True,
    )
    ipa_lemma: Mapped[str] = mapped_column(
        Text,
        nullable=True,
    )

    # -------------------------
    # Definition (gloss) / Explanation / Examples
    # -------------------------
    gloss: Mapped[str] = mapped_column(
        Text,
        nullable=False,
    )
    context_meaning: Mapped[str | None] = mapped_column(
        Text,
        nullable=True,
    )
    explanation: Mapped[str] = mapped_column(
        Text,
        nullable=False,
    )
    examples: Mapped[list[dict]] = mapped_column(
        JSONB,
        nullable=False,
    )

    # -------------------------
    # Source / target languages
    # -------------------------
    src_lang: Mapped[str] = mapped_column(
        Text,
        nullable=False,
    )
    tgt_lang: Mapped[str] = mapped_column(
        Text,
        nullable=False,
    )

    # -------------------------
    # Document -> Word IDs
    # -------------------------
    document_id: Mapped[int] = mapped_column(
        ForeignKey("documents.id"),
        index=True,
        nullable=False,
    )
    page_id: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
    )
    par_id: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
    )
    sent_id: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
    )
    word_id: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
    )

    # -------------------------
    # Sentence / Paragraph of Clicked Word
    # -------------------------
    src_sentence: Mapped[str] = mapped_column(
        Text,
        nullable=False,
    )
    src_paragraph: Mapped[str] = mapped_column(
        Text,
        nullable=False,
    )
    tgt_sentence: Mapped[str] = mapped_column(
        Text,
        nullable=False,
    )
    tgt_paragraph: Mapped[str] = mapped_column(
        Text,
        nullable=False,
    )

    # -------------------------
    # Tokenized alignment snapshot
    # -------------------------
    context_alignment: Mapped[dict[str, Any]] = mapped_column(
        JSONB,
        nullable=False,
    )

    # -------------------------
    # Glossary Item Save Time
    # -------------------------
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )

File: apps/web/app/(protected)/library/feature/types/glossaryItem.ts
import {
  GlossaryItemDefinition,
  GlossaryItemUsage,
  GlossaryItemDefinitionDTO,
  GlossaryItemUsageDTO,
  fromGlossaryContextAlignmentDTO,
} from "@/app/(protected)/documents/feature/types/glossaryItem";

export type LibraryGlossaryItem = {
  glossaryItemId: number;
  documentTitle: string;
  srcLang: string;
  tgtLang: string;
  definition: GlossaryItemDefinition;
  usage: GlossaryItemUsage;
  createdAt: string;
};

export type LibraryGlossaryItemsResponse = {
  glossaryItems: LibraryGlossaryItem[];
};

export type LibraryGlossaryItemDTO = {
  glossary_item_id: number;
  document_title: string;
  src_lang: string;
  tgt_lang: string;
  definition: GlossaryItemDefinitionDTO;
  usage: GlossaryItemUsageDTO;
  created_at: string;
};

export type LibraryGlossaryItemsResponseDTO = {
  glossary_items: LibraryGlossaryItemDTO[];
};

export function fromLibraryGlossaryItemDTO(
  dto: LibraryGlossaryItemDTO
): LibraryGlossaryItem {
  return {
    glossaryItemId: dto.glossary_item_id,
    documentTitle: dto.document_title,
    srcLang: dto.src_lang,
    tgtLang: dto.tgt_lang,
    definition: {
      form: dto.definition.form,
      posForm: dto.definition.pos_form,
      ipaForm: dto.definition.ipa_form,
      lemma: dto.definition.lemma,
      posLemma: dto.definition.pos_lemma,
      ipaLemma: dto.definition.ipa_lemma,
      gloss: dto.definition.gloss,
      srcSentence: dto.definition.src_sentence,
      srcParagraph: dto.definition.src_paragraph,
      tgtSentence: dto.definition.tgt_sentence,
      tgtParagraph: dto.definition.tgt_paragraph,
      contextAlignment: fromGlossaryContextAlignmentDTO(dto.definition.context_alignment),
      documentId: dto.definition.document_id,
      pageId: dto.definition.page_id,
      parId: dto.definition.par_id,
      sentId: dto.definition.sent_id,
      wordId: dto.definition.word_id,
    },
    usage: dto.usage,
    createdAt: dto.created_at,
  };
}

export function fromLibraryGlossaryItemsResponseDTO(
  dto: LibraryGlossaryItemsResponseDTO
): LibraryGlossaryItemsResponse {
  return {
    glossaryItems: dto.glossary_items.map(fromLibraryGlossaryItemDTO),
  };
}
